src/core/lib/event_engine/thread_pool.h

```diff
@@ -47,32 +46,56 @@ class ThreadPool final : public grpc_event_engine::experimental::Forkable {
   void PostforkChild() override;
 
  private:
-  class Thread {
+  class Queue {
    public:
-    explicit Thread(ThreadPool* pool);
-    ~Thread();
+    explicit Queue(int reserve_threads) : reserve_threads_(reserve_threads) {}
+    bool Step();
+    void SetShutdown() { SetState(State::kShutdown); }
+    void SetForking() { SetState(State::kForking); }
+    // Add a callback to the queue.
+    // Return true if we should also spin up a new thread.
+    bool Add(absl::AnyInvocable<void()> callback);
+    void Reset() { SetState(State::kRunning); }
 
    private:
-    ThreadPool* pool_;
-    grpc_core::Thread thd_;
-    void ThreadFunc();
+    enum class State { kRunning, kShutdown, kForking };
+
+    void SetState(State state);
+
+    grpc_core::Mutex mu_;
+    grpc_core::CondVar cv_;
+    std::queue<absl::AnyInvocable<void()>> callbacks_ ABSL_GUARDED_BY(mu_);
+    int threads_waiting_ ABSL_GUARDED_BY(mu_) = 0;
+    const int reserve_threads_;
+    State state_ ABSL_GUARDED_BY(mu_) = State::kRunning;
   };
 
-  void ThreadFunc();
-  void StartNThreadsLocked(int n) ABSL_EXCLUSIVE_LOCKS_REQUIRED(&mu_);
-  static void ReapThreads(std::vector<Thread*>* tlist);
-
-  grpc_core::Mutex mu_;
-  grpc_core::CondVar cv_;
-  grpc_core::CondVar shutdown_cv_;
-  grpc_core::CondVar fork_cv_;
-  bool shutdown_;
-  std::queue<absl::AnyInvocable<void()>> callbacks_;
-  int reserve_threads_;
-  int nthreads_;
-  int threads_waiting_;
-  std::vector<Thread*> dead_threads_;
-  bool forking_;
+  class ThreadCount {
+   public:
+    void Add();
+    void Remove();
+    void BlockUntilThreadCount(int threads, const char* why);
```

Comment:
nit: for C++ification and consistency with new APIs, I'd suggest making this a string_view, even though it's only used currently for its cstring. This will be wanted anyway if/when we move to an alternative logging framework.

Reply:
This would cause an unnecessary allocation for logging and seems worse overall right now.
Revisit when there's an actual advantage?

Reply:
SGTM